File: cpbook/3_problem_solving_paradigms/624_cd.cpp
// UVa 00624 - CD
// https://onlinejudge.org/external/6/624.pdf

#include <iostream>
#include <vector>

using namespace std;

vector<int> tracks, subset, result;
int best, total, n, m;

void search(int k)
{
	if (k == n)	{
		int s = 0;
		vector<int> r;
		for (int i = 0; i < subset.size(); i++)	{
			s += tracks[subset[i]];
			r.push_back(tracks[subset[i]]);
		}
		if (s <= m && s >= best)	{
			best = s;
			result = r;
		}
	}	else	{
		search(k + 1);
		total += tracks[k];
		subset.push_back(k);
		if (total <= m)	{
			search(k + 1);
		}
		subset.pop_back();
		total -= tracks[k];
	}
}


void test_case()
{
	result.clear();
	best = 0;
	total = 0;
	search(0);
	int s = 0;
	for (auto r : result)	{
		cout << r << " ";
		s += r;
	}
	cout << "sum:" << s << "\n";
}

int main()
{
	ios::sync_with_stdio(0);
	cin.tie(0);
	while(cin >> m)	{
		tracks.clear();
		// cin >> m;
		cin >> n;
		for (int i = 0; i < n; i++)	{
			int t;
			cin >> t;
			tracks.push_back(t);
		}
		test_case();
	}
	return 0;
}

File: cpbook/3_problem_solving_paradigms/624_cd.py
# UVa 00624 - CD
# https://onlinejudge.org/external/6/624.pdf

tracks = []
subset = []
result = []
best = 0
m = n = 0

def search(k):
	global subset, tracks, n, m, result, best
	if k == n:
		s = 0
		r = []
		for i in range(len(subset)):
			s += tracks[subset[i]]
			r.append(tracks[subset[i]])
			# print(tracks[subset[i]], "", end="")
		if s <= m:
			if s > best:
				best = s
				result = r[:]
			# print(r, end="")
			# print("sum = ", s)
	else:
		search(k + 1)
		subset.append(k)
		search(k + 1)
		subset.pop()



def test_case(line):
	global tracks, n, m, result
	info = [int(x) for x in line.split()]
	m = info[0]
	n = info[1]
	tracks = info[2:]
	print(tracks)
	result = []
	search(0)
	for r in result:
		print(r, "", end="")
	print("sum:", sum(result), sep="")


if __name__ == "__main__":
	while True:
		try:
			line = input()
		except EOFError:
			break
		if line == "":
			break
		test_case(line)

